apps/emqx_mix_utils/lib/mix/tasks/compile.asn1.ex

```diff
@@ -0,0 +1,82 @@
+defmodule Mix.Tasks.Compile.Asn1 do
+  use Mix.Task.Compiler
+
+  @recursive true
+  @manifest_vsn 1
+  @manifest "compile.asn1"
+  # TODO: use manifest to track generated files?
+
+  @impl true
+  def manifests(), do: [manifest()]
+  defp manifest(), do: Path.join(Mix.Project.manifest_path(), @manifest)
+
+  @impl true
+  def run(_args) do
+    add_to_path_and_cache(:asn1)
+
+    Mix.Project.get!()
+    config = Mix.Project.config()
+    app_root = File.cwd!()
+
+    asn1_srcs = config[:asn1_srcs] || []
+    asn1_opts = config[:asn1_opts] || %{}
+    out_dir = Map.fetch!(asn1_opts, :out_dir)
+    opts = [:per, :noobj, outdir: to_charlist(out_dir)]
```

Comment:
Compiler options are the only important part, yet they're hidden in the middle of some function. They should be at least moved to a visible place.
Also, it hardcodes PER encoding for all asn.1 files, which is too limiting. 

Can we just call `erlc` via shell? Why do we need a wrapper that essentially does the it, but in 80 lines?

Reply:
Since the current solution [already uses hard-coded flags](https://github.com/emqx/emqx/blob/ffa69df6f8fc01317b986de223744fdb4cf07343/apps/emqx_durable_storage/rebar.config#L7) and command to compile, I don't think there's much difference nor need to make it generic or different.  We just use this to compile only EMQX project.  As new uses cases emerge, we can simply update the command here.

Anyways, I've moved the flags to being an option exposed in `mix.exs`, just like what's now with `rebar.config`.

Reply:
> Since the current solution [already uses hard-coded flags](https://github.com/emqx/emqx/blob/ffa69df6f8fc01317b986de223744fdb4cf07343/apps/emqx_durable_storage/rebar.config#L7) and command to compile

There's a big difference: "current solution" is simply a hardcoded hook _in the application that defines a consumes the schema_. It's not a _plugin_, it's local and not generic. 

This file is named `apps/emqx_mix_utils/lib/mix/tasks/compile.asn1.ex`, i.e. it's a generic util for compiling ASN.1. It's not part of durable_storage, so if anyone else uses it and needs e.g. BER, they change the flag and it changes all encoders and in the worst case old data becomes unreadable. 

So I ask to move this compile hook back to durable_storage. We don't need to create and maintain a generic ASN.1 mix plugin.

Reply:
The compile options have been moved to `emqx_durable_storage/mix.exs`, so they are more explicit.  

Also, again, it's as generic as our build pipeline needs it to be, since this `emqx_mix_utils` app is only used for mix utilities and not released anywhere.  Currently, it's sufficient to build our project.  If we need to expand it at a later time, we can.